ts_src/taproot.ts

```diff
@@ -0,0 +1,205 @@
+import assert = require('assert');
+import FastPriorityQueue = require('fastpriorityqueue');
+import * as bcrypto from './crypto';
+const ecc = require('tiny-secp256k1');
+
+// const SECP256K1_ORDER = Buffer.from('fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141', 'hex')
+const INITIAL_TAPSCRIPT_VERSION = Buffer.from('c0', 'hex');
+
+const TAPLEAF_TAGGED_HASH = bcrypto.sha256(Buffer.from('TapLeaf'));
+const TAPBRANCH_TAGGED_HASH = bcrypto.sha256(Buffer.from('TapBranch'));
+const TAPTWEAK_TAGGED_HASH = bcrypto.sha256(Buffer.from('TapTweak'));
+
+interface WeightedTapScript {
+  /** A TapLeaf or TapBranch tagged hash */
+  taggedHash: Buffer;
+  weight: number;
+}
+
+/**
+ * Trims the leading 02/03 byte from an ECDSA pub key to get a 32 byte schnorr
+ * pub key with x-only coordinates.
+ * @param pubkey A 33 byte pubkey representing an EC point
+ * @returns a 32 byte x-only coordinate
+ */
+export function trimFirstByte(pubkey: Buffer): Buffer {
+  assert(pubkey.length === 33);
+  return pubkey.slice(1, 33);
+}
+
+/**
+ * Aggregates a list of public keys into a single MuSig public key
+ * according to the MuSig paper.
+ * @param pubkeys The list of pub keys to aggregate
+ * @returns a 32 byte Buffer representing the aggregate key
+ */
+export function aggregateMuSigPubkeys(pubkeys: Buffer[]): Buffer {
+  // sort keys in ascending order
+  pubkeys.sort();
+
+  const trimmedPubkeys: Buffer[] = [];
+  pubkeys.forEach(pubkey => {
+    const trimmedPubkey = trimFirstByte(pubkey);
+    trimmedPubkeys.push(trimmedPubkey);
+  });
```

Comment:
Shouldn't these be sorted after trimming?

Reply:
Also, shouldn't there be a check (and throw an error) if two pubkeys are the same?

Reply:
(for p2ms we allow multiple of the same pubkey since Bitcoin allows this... but for aggregating, I think two of the same pubkey is not mathematically possible? I could be wrong.)

Reply:
You're correct about sorting after trimming. I think a sanity check here to prevent duplicate pubkeys makes sense even if it's technically possible.